keras/02-the-layer-class.md

```diff
@@ -0,0 +1,317 @@
+something is called and don't want it to be called multiple times, you should
+use this pattern. 
+
+
+### The `Layer.add_weight()` function
+
+([Source](https://github.com/keras-team/keras/blob/v2.6.0/keras/engine/base_layer.py#L528))
+
+We would also like to see how these `tf.Variable`s are created in the
+`add_weight()` function. Here is the pseudo-code for the core logic of the
+`add_weight()` function.
+
+It creates the variable and asks the backend to track the variable. The
+variable will be appended to different lists depending on if it is trainable.
+
+The [backend](https://github.com/keras-team/keras/blob/v2.6.0/keras/backend.py)
+is a file containing some abstractions of the tensorflow functionalities. In
+many cases, the Keras code would interact with the backend instead of directly
+calling the TensorFlow APIs.
+
+([Source](https://github.com/keras-team/keras/blob/v2.6.0/keras/engine/base_layer.py#L528))
+
+```py
+class Layer(module.Module, ...):
+  def add_weight(self, ...):
+    ...
+    variable = ...  # Create the variable.
+    backend.track_variable(variable)  # Track the variable.
+    if trainable:
+      self._trainable_weights.append(variable)
+    else:
+      self._non_trainable_weights.append(variable)
+```
+
+The process for creating the variable is a [function
+call](https://github.com/keras-team/keras/blob/v2.6.0/keras/engine/base_layer.py#L647),
+which is not so different from using `tf.Variable(...)` to directly create the
+variable.
+
+We need the backend to track the variable for model saving and computation
+optimization. Now, the question is how the backend is tracking the variable.
+The code is shown as follows.
+
+([Source](https://github.com/keras-team/keras/blob/v2.6.0/keras/backend.py#L1072))
+
+```py
+def track_variable(v):
+  """Tracks the given variable for initialization."""
+  if context.executing_eagerly():
+    return
+  graph = v.graph if hasattr(v, 'graph') else get_graph()
+  _GRAPH_VARIABLES[graph].add(v)
+```
+
+We encountered two important concepts: [eager
+mode](https://www.tensorflow.org/guide/eager) and [graph
+mode](https://www.tensorflow.org/guide/intro_to_graphs). You can click the
+link for detailed introductions.
+
+Here is a short explanation. You can think of eager execution as plain Python
+code execution. The tensors are all concrete values instead of placeholders.
+The operations are read and executed only when we run that line of code in the
+Python interpreter.
+
+However, in graph mode, all the tensors and operations are collected in advance
+to build the computational graph before any actual value is input for
+computation. The graph is then optimized for execution speed. It is similar to
+a compiled language, like the C programming language, which you can turn on
+various optimization options to make the compiled executable file run faster.
+
+> **_TensorFlow API_** <br>
+`tf.executing_eagerly()` is to check whether TensorFlow is running in eager
```

Comment:
In keras code, we usually don't use executing_eagerly to do the check, since it will return false when code is in the body of tf.function (which is mostly the case). We use tf.compat.v1.executing_eagerly_outside_functions to the actual check of whether eager is enabled or not.